Add CSS Color 4 syntax support - integration with changes from maplibre-style

kajkal opened the pull request:
Integration with changes from PR [#94](https://github.com/maplibre/maplibre-style/pull/94) from [maplibre-style](https://github.com/maplibre/maplibre-style).

- update color-related integration tests
- add color-interpolation render tests

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

kajkal commented on test/integration/expression/expression.test.ts:
```diff
@@ -1,6 +1,7 @@
 import path from 'path';
 import fs from 'fs';
 import glob from 'glob';
+import {printDiffOrStringify} from 'jest-matcher-utils';
```
Anticipating potential questions: library `jest-matcher-utils` is already part of the project because it is one of the dependencies of `jest`, the new import in `package.json` is only for the ability to import a specific function in the code. The mentioned method is `jest` agnostic, it does not need the `jest` environment to work properly therefore it will not block a potential migration from `jest` to another testing framework.

Now when many tests from the `expression` suite fail it is difficult to assign the expected/actual to a specific test because they are displayed with console.log and, with many tests, they are easily mixed in with others (see [current_output.txt](https://github.com/maplibre/maplibre-gl-js/files/11195894/current_output.txt) file). With this change, the difference between expected/actual is displayed along with information about the failure of a particular test (see [new_output.txt](https://github.com/maplibre/maplibre-gl-js/files/11195891/new_output.txt) file).

HarelM replied:
The problem is that I'm already very much burden with updating dependencies, so I would like to avoid adding more work in that area.
If this is only for printing failing tests, I would rather avoid adding a dependency only for that, ATM.
My two cents at least...

kajkal replied:
I understand your concerns however this library is from under the `jest` umbrella and has the same model of releasing new versions. Its updates will take place together with updates of other libraries related to `jest`:
```json
"@types/jest": "^29.5.0",
"jest": "^29.5.0",
"jest-environment-jsdom": "^29.5.0",
"jest-matcher-utils": "^29.5.0",
```
So I think that one more `jest` library should not provide many new problems :)
Its use allows a quick and visually appealing (formatting, colors, etc.) assessment of what's wrong with a given test, which significantly contributes to the positive developer experience.

However, these are just my two cents added to this project, and if you still think this change is unnecessary, I will not insist on it.

HarelM replied:
I tried to reduce the maintainance burden but failed, so for now, only critical libraries should be added, sorry...

kajkal replied:
As you wish, removed.

HarelM commented:
@kajkal can you try and update the branch here and let me know if everything is working as expected in terms of updating the spec version?

kajkal commented:
Updated, colors (parsing/interpolation) are as expected, all tests are passing - it seems fine to me 👍

HarelM commented:
Any chance for a render test? So that this change could be visually seen?

kajkal commented:
Oh ok, those exist, I completely missed them and was only looking at the results of `jest` 😩

drwestco commented on src/style/properties.ts:
```diff
@@ -483,9 +483,10 @@ export class DataConstantProperty<T> implements Property<T, T> {
-        const interp: any = interpolateFactory(this.specification.type as any);
-        if (interp) {
-            return interp(a, b, t);
+        const interpolationType = this.specification.type as 'number' | 'color' | 'array' | 'padding';
```
There should be a type for this in maplibre-gl-style-spec instead of hard-coding the possible values here.
```TypeScript
type interpolateType = keyof typeof interpolates;

kajkal replied:
Yes, you are right. This change would now require a separate PR to maplibre-gl-style-spec because I read your comment too late, so I'm not sure if it's worth it for those 2 repetitions - what do you think?

HarelM replied:
I think it's worth it as I haven't released a new version yet...

drwestco commented on src/style/properties.ts:
```diff
@@ -542,9 +543,11 @@ export class DataDrivenProperty<T> implements Property<T, PossiblyEvaluatedPrope
-        if (interp) {
-            return new PossiblyEvaluatedPropertyValue(this, {kind: 'constant', value: interp(a.value.value, b.value.value, t)}, a.parameters);
+        const interpolationType = this.specification.type as 'number' | 'color' | 'array' | 'padding';
+        const interpolationFn = interpolates[interpolationType] as ((from: T, to: T, t: number) => T) | undefined;
```
`interpolates` should have a proper type signature, so casting via `as` shouldn't be necessary.

kajkal replied:
I also think that `interpolates` should have a proper type signature in this case, but unfortunately sometimes TS doesn't want to cooperate :D
See [ts playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABGEBbARgUwE4ApjZyoBcyaW2ANIlHKShjtVPeTgJSuPaIDeAUIiGJsmKCGxICRRAGoaiAFSJctRAFpE01OwDc-AL79+oSLASIAhtmyWAnvkIky3ANoBdZnRcUPzLhScPjgefILCouKSWk4AdKiWAA64uAAm1DDsiAC8AHzBeOk0cK4wnjTseobGEAgAzlCIMGBQOIlwADaWrXU5fAwU1Na2dgb6-FB2iZiIAMqTHZgACoTT2JOz0xAwwDAQ3fBgACpTM9mIAOQDOBeIAD6XDdjNAOa3DxfocJ2YlmDvl0wDABF1qHTg2BBw3sIMSllSqVeF3GEC6dV6ABFupYAML1KB-KArOBrSYAHiO+QEwhop1I8zsi2JpLsm0w212+3Mx1O+hptTATxA0AhqjpcwWy1WOA2Wx2ewOCBO0yy1JpwigAAsYHVYpNpn19Zg+cIjDTmq1sO0uq1cJZSEdqOgHf4CkEjmF1cIBQ0mi02p1FTyDectTq9acrL0rmxIfdLtC7Mjwl6fY0LQGbYcAGJIc4Zq2Bnqlf2FrNK07uE1enYqAvWoO51Upr0RMQSJD1os5sB2p3MKqtxAGRCYDp1GZqodCSIdqzV9Vm03VIA).

kajkal commented:
As some render tests proved, there are cases in which an already parsed instance of `Color` class is passed to `Color.parse` function, contrary to [my assumptions](https://github.com/maplibre/maplibre-style-spec/pull/94#discussion_r1161059154).
I created PR [#128](https://github.com/maplibre/maplibre-style-spec/pull/128) (blocker for this PR) to fix that.

I added 2 render tests for color interpolation and added the ability to change the default sensitivity to color differences.
The tests in the previous configuration checked colors with a approximation: `rgb(46,215,46)` vs `rgb(0,255,0)` => pass.
By changing the sensitivity for tests directly related to colors, they are now able to detect a small change in the value of even one of the sRGB channels: `rgb(28,199,255)` vs `rgb(28,200,255)` => fail.

drwestco commented:
You can use the 'keyof typeof' syntax here instead of string literals, so
it keeps working when new interpolatable types are added. If we run into
more cases needing the interpolationType, we can then consider a change in
the style spec repo.

On Sun, Apr 16, 2023, 2:19 PM Karol Leśniak ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/style/properties.ts
> <https://github.com/maplibre/maplibre-gl-js/pull/2376#discussion_r1168026496>
> :
>
> > @@ -483,9 +483,10 @@ export class DataConstantProperty<T> implements Property<T, T> {
>      }
>
>      interpolate(a: T, b: T, t: number): T {
> -        const interp: any = interpolateFactory(this.specification.type as any);
> -        if (interp) {
> -            return interp(a, b, t);
> +        const interpolationType = this.specification.type as 'number' | 'color' | 'array' | 'padding';
>
> Yes, you are right. This change would now require a separate PR to
> maplibre-gl-style-spec because I read your comment too late, so I'm not
> sure if it's worth it for those 2 repetitions - what do you think?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/maplibre/maplibre-gl-js/pull/2376#discussion_r1168026496>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFVM5SY75IHJBZLUYRC4YDXBRO6NANCNFSM6AAAAAAWZVR43Y>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

HarelM commented on test/integration/render/tests/background-color/colorSpace-hcl/style.json:
```diff
@@ -3,7 +3,8 @@
       "width": 64,
-      "height": 64
+      "height": 64,
+      "threshold": 0.001
```
Can we use an image that doesn't need this threshold change? I've struggled in the past to find this override, and I would like to avoid it as much as possible...

kajkal replied:
Now I see, I checked the results on Windows and on Ubuntu 22.04 (WSL), there the tests passed, here they have problems. 

I changed it after I decided that with their current threshold tests do not quite reliably check the colors, they allow too much margin, as I wrote above ^

> The tests in the previous configuration checked colors with a approximation: `rgb(46,215,46)` vs `rgb(0,255,0)` => pass. By changing the sensitivity for tests directly related to colors, they are now able to detect a small change in the value of even one of the sRGB channels: `rgb(28,199,255)` vs `rgb(28,200,255)` => fail.

Well, I will restore the previous threshold.

HarelM replied:
My bad, I missed that comment. Can we tighten the threshold for all the tests? Will they still pass?

kajkal replied:
Unfortunately, no. More than 100 tests don't pass then, and I currently lack the knowledge of GL rendering differences between platforms, the Ubuntu machine at hand, and the time to address this.

HarelM commented on test/integration/render/tests/regressions/mapbox-gl-js#3107/style.json:
```diff
@@ -45,8 +45,8 @@
       "paint": {
-          "fill-outline-color": "hsla(0, 0, 0, 0)",
-          "fill-color": "hsla(0, 0, 0, 0)",
+          "fill-outline-color": "hsla(0, 0%, 0%, 0)",
```
Is this mandatory now (the extra "%")? or will this work with the previous notation?

HarelM replied:
If this is the case we need to make sure we are documenting this more clearly in the changelog.

kajkal replied:
Hmm... `hsl` without `%` is not a valid CSS color, Maplibre should not have supported it previously and there is nothing in the [documentation](https://maplibre.org/maplibre-style-spec/types/#color) about exceptional treatment of `hsl` without percentages. 
What changelog entry do you have in mind in this case?

HarelM replied:
In the same chengelog entry here we should add that "if you have an hcl color without % you should add it to create a valid one" or something similar - a kind of migration guide.

kajkal replied:
Ok, so for `hsl` without `%` - breaking change. And what to do with the changelog entries added to maplibre-style-spec - repeat them in this repo changelog too?

HarelM replied:
Just a single entry with the breaking change, migration guide and short description of the new functionality here.

HarelM commented:
Seems like there's a render test that is failing. Note that I have recently added some instructions on how to get an image from the CI to get the test passing. There is a report that should be attached to the CI run, you can get it by going to the actions tab here in github, I believe.

kajkal commented:
Thanks, I was not aware that this was possible!
I adjusted the threshold for my new color interpolation tests - It should be working now.

HarelM commented:
Generally approved, you'll need to set it for ready for review and resolve conflicts to have it merged.
